src/LndClient/Data/LndEnv.hs

```diff
@@ -84,30 +90,40 @@ newLndEnv ::
   LndHexMacaroon ->
   LndHost ->
   LndPort ->
-  Either String LndEnv
+  Either LndError LndEnv
 newLndEnv pwd cert mac host port =
-  --
-  -- TODO : implement smart constructors
-  -- verify certificate and all other data validity
-  -- host, port etc
-  --
-  Right $
-    LndEnv
-      { envLndWalletPassword = pwd,
-        envLndHexMacaroon = mac,
-        envLndLogErrors = True,
-        envLndGrpcConfig =
-          ClientConfig
-            { clientServerHost = Host $ coerce host,
-              clientServerPort = Port $ coerce port,
-              clientArgs = [],
-              clientSSLConfig =
-                Just $
-                  ClientSSLConfig
-                    { serverRootCert = Just $ coerce cert,
-                      clientSSLKeyCertPair = Nothing,
-                      clientMetadataPlugin = Nothing
-                    },
-              clientAuthority = Nothing
-            }
-      }
+  createLndEnv <$> validatePort port <*> validateCert cert
+  where
+    validatePort :: LndPort -> Either LndError Int
+    validatePort p = do
+      let port32 :: Word32 = coerce p
+      let maybePort :: Maybe Int = U.safeFromIntegral port32
+      maybeToRight (LndEnvError "Wrong port") maybePort
+    validateCert :: LndTlsCert -> Either LndError LndTlsCert
+    validateCert c = do
+      pemsM <- first (LndEnvError . LT.pack) $ Pem.pemParseBS $ coerce c
+      pem <- note (LndEnvError $ LT.pack "No pem found") $ safeHead pemsM
+      case decodeSignedCertificate $ Pem.pemContent pem of
```

Comment:
Expressions like `Left _ -> Left ...` and `Right _ -> Right ...` are a bit out of functional programming paradigm. Nothing wrong with it, it works for sure, just conceptually in functional programming boilerplate `case` expressions like this are considered too low-level. My personal experience says me to avoid `if` and `case` if possible and replace with more generic functions (related to monads, functors applicatives and so on). In this particular place I propose to use bifunctor nature of Either. Something like this

```haskell
bimap 
  (LndEnvError . LT.pack . ("Certificate is not valid: " <>))
  (const c)
  (decodeSignedCertificate $ Pem.pemContent pem)
```

Reply:
 this is good one